src/main/kotlin/persistence/repo/UserRepo.kt

```diff
@@ -0,0 +1,143 @@
+
+    /**
+     * function to create and insert a user into the database
+     * takes in a audioHash and a path to a recording to creaete
+     */
+    override fun insert(user: User): Observable<Int> {
+        // creates observable to return generated int
+        return Observable.fromCallable {
+            userMapper.mapToEntity(Observable.just(user))
+        }.flatMap {
+            it
+        }.map {
+            userEntityDao.insert(it)
+            userEntityDao.fetchByAudiohash(it.audiohash).first()
+        }.flatMap {
+            user.id = it.id
+            updateUserLanguageReferences(user, it.id)
+            userPreferencesMapper.mapToEntity(Observable.just(user.userPreferences))
+        }.map {
+            userPreferencesEntityDao.insert(it)
+            user.id
+        }.subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * gets user by Id
+     */
+    override fun getById(id:Int): Observable<User> {
+        return userMapper.mapFromEntity(Observable.fromCallable {
+            userEntityDao.fetchById(id).first()
+        }).subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * gets all the users currently stored in db
+     */
+    override fun getAll(): Observable<List<User>> {
+        return Observable.fromCallable {
+            userEntityDao.findAll().toList()
+        }.flatMap {
+            val userList = it.map { userMapper.mapFromEntity(Observable.just(it)) }
+            Observable.zip(userList) {
+                it.toList() as List<User>
+            }
+        }.subscribeOn(Schedulers.io())
+    }
+
+
+    override fun update(user: User): Completable {
+        return Completable.fromObservable(
+            userMapper.mapToEntity(Observable.just(user)).flatMap {
+                userEntityDao.update(it)
+                updateUserLanguageReferences(user, user.id)
+                userPreferencesMapper.mapToEntity(Observable.just(user.userPreferences))
+            }.doOnNext {
+                userPreferencesEntityDao.update(it)
+            }).subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * deletes user by id
+     */
+    override fun delete(user: User): Completable {
+        return Completable.fromObservable(
+            userMapper.mapToEntity(Observable.just(user)).map {
+                userEntityDao.delete(it)
+            }).subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * This function should be unnecessary because we have add and remove function
+     * this then enforces clients to add and remove languages through the database
+     * keeping just in case we want this
+     */
+
+    private fun updateUserLanguageReferences(user: User, userId: Int) {
+        // inserts source and target languages into user language relationship table
+        val newSourceUserLanguages = user.sourceLanguages.map {
+            UserLanguagesEntity(
+                    userId,
+                    it.id,
+                    1
```

Comment:
what is 1?

Reply:
since in sqlite we cant store boolean values for isSource, so we're using integer values, so right now 1 is true and 0 is false

Reply:
We are creating a `UserLanguageEntity` for a source language here. Since SQLite doesn't have a native boolean, 1 initializes the isSource field to true.